src/dust/coagulation/coagulation.cpp

```diff
@@ -0,0 +1,308 @@
+  const Real s_max = pin->GetReal("dust", "max_size");
+  const Real mmin = 4.0 * M_PI / 3.0 * rho_p * std::pow(s_min, 3);
+  const Real mmax = 4.0 * M_PI / 3.0 * rho_p * std::pow(s_max, 3);
+  const Real cond = 1.0 / (1.0 - nm) * std::log(mmin / mmax);
+  const Real conc = std::log(mmin);
+  if (std::exp(cond) > std::sqrt(2.0)) {
+    std::stringstream msg;
+    msg << "### FATAL ERROR in dust with coagulation: using nspecies >"
+        << std::log(mmax / mmin) / (std::log(std::sqrt(2.0))) + 1. << " instead of " << nm
+        << std::endl;
+    PARTHENON_FAIL(msg);
+  }
+
+  ParArray1D<Real> dust_size("dustSize", nm);
+  auto dust_size_host = dust_size.GetHostMirror();
+
+  for (int n = 0; n < nm; ++n) {
+    Real mgrid = std::exp(conc + cond * n);
+    dust_size_host(n) = std::pow(3.0 * mgrid / (4.0 * M_PI * cpars.rho_p), 1. / 3.);
+  }
+  dust_size.DeepCopy(dust_size_host);
+
+  // allocate array and assign values
+  cpars.klf = ParArray2D<int>("klf", nm, nm);
+  cpars.mass_grid = ParArray1D<Real>("mass_grid", nm);
+  const int n2DRv = coag2DRv::last2;
+  cpars.coagR3D = ParArray3D<Real>("coagReal3D", n2DRv, nm, nm);
+  cpars.cpod_notzero = ParArray3D<int>("idx_nzcpod", nm, nm, 4);
+  cpars.cpod_short = ParArray3D<Real>("nzcpod", nm, nm, 4);
+
+  if (Dust::cgsunit == NULL) {
+    Dust::cgsunit = new CGSUnit;
+  }
+  if (!Dust::cgsunit->isSet()) {
+    Dust::cgsunit->SetCGSUnit(pin);
+  }
+  const Real dfloor = pin->GetOrAddReal("dust", "dfloor", 1.0e-25);
+  // convert to CGS unit
+  cpars.dfloor = dfloor * Dust::cgsunit->mass0 / Dust::cgsunit->vol0;
+  Real a = std::log10(mmin / mmax) / static_cast<Real>(1 - nm);
+
+  initializeArray(nm, cpars.pGrid, cpars.rho_p, cpars.chi, a, dust_size, cpars.klf,
+                  cpars.mass_grid, cpars.coagR3D, cpars.cpod_notzero, cpars.cpod_short);
+
+  params.Add("coag_pars", cpars);
+
+  // other parameters for coagulation
+  const int nstep1Coag = pin->GetOrAddReal("problem", "nstep1Coag", 50);
+  params.Add("nstep1Coag", nstep1Coag);
+  Real dtCoag = 0.0;
+  params.Add("dtCoag", dtCoag, Params::Mutability::Restart);
+  const Real alpha = pin->GetOrAddReal("dust", "coag_alpha", 1.e-3);
+  params.Add("coag_alpha", alpha);
+  const int scr_level = pin->GetOrAddReal("dust", "coag_scr_level", 0);
+  params.Add("coag_scr_level", scr_level);
+
+  return coag;
+}
+
+// using Kokkos par_for() loop
+void initializeArray(const int nm, int &pGrid, const Real &rho_p, const Real &chi,
+                     const Real &a, const ParArray1D<Real> dsize, ParArray2D<int> klf,
+                     ParArray1D<Real> mass_grid, ParArray3D<Real> coag3D,
+                     ParArray3D<int> cpod_notzero, ParArray3D<Real> cpod_short) {
+
+  int ikdelta = coag2DRv::kdelta;
+  auto kdelta = Kokkos::subview(coag3D, ikdelta, Kokkos::ALL, Kokkos::ALL);
+  int icoef_fett = coag2DRv::coef_fett;
+  auto coef_fett = Kokkos::subview(coag3D, icoef_fett, Kokkos::ALL, Kokkos::ALL);
```

Comment:
I would prefer to not have raw Kokkos calls like this. It should be just as easy to index `coag3D` by `coag2DRv::kdelta` index in the kernel.

Reply:
I do not know how to use the index to reduce 3D ParArray to 2D ParArray. The following does not work:
ParArray2D<Real> &kdelta = coag3D(ikdelta);

Reply:
You index it in the kernel, not outside it. `coag3d(b, ikdelta, i,j)` instead of `kdelta(i,j)`